Topic/virt x86: memory hotplug support with GED

yangzhon opened the pull request:
@sameo @mcastelino @rbradford @liujing2 

Like CPU acpi hotplug did, we add the memory hotplug with GED solution.  This PR is in WIP, and please help review. thanks!

yangzhon committed:
i386: virt: Hook the virt memory hotplug into ACPI

This patch add the virt memory hotplug/unplug handler into
the ACPI framework.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: acpi: Implement ACPI memory hotplug hooks

We register acpi device with memory hotplug framework and then
implement the hotplug/unplug device hooks.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: acpi: add memory hotplug support in dsdt

We add the memory hotplug in dsdt acpi table and use GED
events to handle memory hotplug.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@sameo @mcastelino @rbradford @liujing2

rebased on #37 , thanks!

sameo commented:
@yangzhon When I rebase your PR on top of PR #39 and then add the following patch:

```C
diff --git a/hw/i386/virt/acpi.c b/hw/i386/virt/acpi.c
index 65f9dde31e..6a1c08553c 100644
--- a/hw/i386/virt/acpi.c
+++ b/hw/i386/virt/acpi.c
@@ -100,6 +100,9 @@ static void virt_send_ged(AcpiDeviceIf *adev, AcpiEventStatusBits ev)
     if (ev & ACPI_CPU_HOTPLUG_STATUS) {
         /* We inject the CPU hotplug interrupt */
         qemu_irq_pulse(s->gsi[VIRT_GED_CPU_HOTPLUG_IRQ]);
+    } else if (ev & ACPI_MEMORY_HOTPLUG_STATUS) {
+        /* We inject the memory hotplug interrupt */
+        qemu_irq_pulse(s->gsi[VIRT_GED_MEMORY_HOTPLUG_IRQ]);
     }
 }

```

memory hotplug works for me.

sameo commented on hw/acpi/reduced.c:
```diff
@@ -46,7 +46,16 @@
+{
+    uint32_t nr_mem = ms->ram_slots;
+
+    build_memory_hotplug_aml(dsdt, nr_mem, "\\_SB.PCI1", GED_DEVICE);
```
With PR #39 the last argument here should be `NULL`.

yangzhon replied:
Done, thanks!

sameo commented on hw/i386/virt/memory.c:
```diff
@@ -56,9 +58,46 @@ MemoryRegion *virt_memory_init(VirtMachineState *vms)
         sysfw_firmware_init(system_memory, false);
     }
-    
+
```
Nitpick: Unneeded change.

yangzhon replied:
Done, thanks!

sameo commented on include/hw/acpi/pc-hotplug.h:
```diff
@@ -31,5 +31,6 @@
 #define CPU_HOTPLUG_RESOURCE_DEVICE PRES
 
 #define ACPI_MEMORY_HOTPLUG_BASE 0x0a00
+#define VIRT_MEMORY_HOTPLUG_BASE 0x0a00
```
Let's simply re-use `ACPI_MEMORY_HOTPLUG_BASE` here.

yangzhon replied:
Done, thanks!

sameo commented on hw/i386/virt/memory.c:
```diff
@@ -32,7 +32,7 @@ MemoryRegion *virt_memory_init(VirtMachineState *vms)
     MemoryRegion *ram = g_new(MemoryRegion, 1), *lowmem, *highmem, *system_memory = get_system_memory();
     uint64_t lowmem_size, highmem_size;
-    
+
```
Unneeded change.

yangzhon replied:
Done, thanks!

sameo commented on hw/pci-host/pci_lite.c:
```diff
@@ -65,6 +66,25 @@ typedef struct PCILiteHost {
+        hole64_start = vms->hotplug_memory.base;
+        hole64_start += memory_region_size(&vms->hotplug_memory.mr);
+    } else {
+        hole64_start = 0x100000000ULL + vms->above_4g_mem_size;
```
Can we please create a `#define` for this `0x100000000ULL` value?

sameo commented on hw/acpi/reduced.c:
```diff
@@ -34,7 +34,7 @@
 #include "hw/acpi/reduced.h"
 #include "qemu/range.h"
 #include "hw/nvram/fw_cfg.h"
-
```
Please keep this space, I prefer when `#include` are clearly separated. I'm a little picky, sorry.

yangzhon replied:
Done, thanks!

sameo reviewed (changes requested):
Really minor stuff, except for the missing `virt_send_ged()` part.
Clean work overall, thanks a lot.

yangzhon committed:
virt: pci: change 64bit pci hole for memory hotplug

The previous 64bit pci hole used fixed value, the new implementation
add the memory hotplug.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
i386: virt: initialize hotplug memory region

Initialize hotplug memory address space for memory
hotplug usage.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@sameo @mcastelino @rbradford @liujing2

Rebased on #39 and addressed samuel's comments, thanks!

sameo commented:
LGTM

sameo commented:
The x86 CI is failing when testing the virt machine type. Manual testing also fails.
This command:

```
sudo -E ../../build-x86-64/x86_64-softmmu/qemu-system-x86_64 -machine virt,accel=kvm,kernel_irqchip -pidfile qemu.pid -qmp unix:monitor.sock,server,nowait -bios /home/samuel/workloads//OVMF.fd -smp 2,cores=1,threads=1,sockets=2,maxcpus=32 -m 512,slots=4,maxmem=16384M -cpu host -nographic -no-user-config -nodefaults -daemonize -drive file=testvm.img,if=none,id=drive-virtio-disk0,format=qcow2 -device virtio-blk-pci,scsi=off,drive=drive-virtio-disk0,id=virtio-disk0,bootindex=1 -device sysbus-debugcon,iobase=0x402,chardev=debugcon -chardev file,path=/tmp/debug-log,id=debugcon -device virtio-blk-pci,drive=cloud -drive if=none,id=cloud,file=seed.img,format=raw -netdev user,id=mynet0,hostfwd=tcp::2222-:22,hostname=nemuvm -device virtio-net-pci,netdev=mynet0 -drive file=testscsi.img,if=none,id=drive-virtio-disk1,format=raw -device virtio-scsi-pci,id=virtio-disk1 -object memory-backend-file,id=mem0,share,mem-path=testnvdimm.img,size=51200 -device nvdimm,memdev=mem0,id=nv0 -device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=3 -device virtio-serial-pci,id=virtio-serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev socket,id=charconsole0,path=console.sock,server,nowait -netdev tap,fd=3,id=hostnet0,vhost=on,vhostfd=4 -device virtio-net-pci,netdev=hostnet0,id=net0,mac=5a:53:24:60:25:ae -device virtio-rng-pci,rng=rng0 -object rng-random,filename=/dev/random,id=rng0 -device virtio-balloon-pci -object cryptodev-backend-builtin,id=cryptodev0 -device virtio-crypto-pci,id=crypto0,cryptodev=cryptodev0
```

fails with this PR.

yangzhon commented:
@sameo @mcastelino @rbradford

What's the error found in your side? I justly tested your CI command with or without this PR, the error info are same. Below are error info when boot guest with this CI command
qemu-system-x86_64: -netdev tap,fd=3,id=hostnet0,vhost=on,vhostfd=4: TUNGETIFF ioctl() failed: Inappropriate ioctl for device
TUNSETOFFLOAD ioctl() failed: Inappropriate ioctl for device
qemu-system-x86_64: -netdev tap,fd=3,id=hostnet0,vhost=on,vhostfd=4: vhost_set_owner failed: Inappropriate ioctl for device (25)
qemu-system-x86_64: -netdev tap,fd=3,id=hostnet0,vhost=on,vhostfd=4: warning: vhost-net requested but could not be initialized

rbradford commented:
@yangzhon You need to make sure your user has access to /dev/vhost-net. The simplest way is to run as root.

rbradford commented:
@yangzhon I was able to reproduce this under gdb: here is the BT, i'll look into the bug further:

```
Thread 1 "qemu-system-x86" received signal SIGSEGV, Segmentation fault.
memory_region_size (mr=mr@entry=0x160) at /home/rob/src/nemu/memory.c:1736
1736	    if (int128_eq(mr->size, int128_2_64())) {
(gdb) bt
#0  memory_region_size (mr=mr@entry=0x160) at /home/rob/src/nemu/memory.c:1736
#1  0x0000555555a2f10f in pc_dimm_memory_plug (dev=dev@entry=0x55555706c5f0, hpms=hpms@entry=0x150, mr=mr@entry=0x55555706c690, align=4096, errp=errp@entry=0x7fffffffd740)
    at /home/rob/src/nemu/hw/mem/pc-dimm.c:56
#2  0x00005555558e5243 in virt_dimm_plug (errp=0x7fffffffd7a0, dev=0x55555706c5f0, hotplug_dev=<optimised out>) at /home/rob/src/nemu/hw/i386/virt/virt.c:567
#3  virt_machine_device_plug_cb (hotplug_dev=<optimised out>, dev=0x55555706c5f0, errp=0x7fffffffd7a0) at /home/rob/src/nemu/hw/i386/virt/virt.c:638
#4  0x00005555559d7983 in device_set_realized (obj=<optimised out>, value=<optimised out>, errp=0x7fffffffd8c8) at /home/rob/src/nemu/hw/core/qdev.c:862
#5  0x0000555555b40ace in property_set_bool (obj=0x55555706c5f0, v=<optimised out>, name=<optimised out>, opaque=0x55555705b490, errp=0x7fffffffd8c8) at /home/rob/src/nemu/qom/object.c:1923
#6  0x0000555555b44c7f in object_property_set_qobject (obj=obj@entry=0x55555706c5f0, value=value@entry=0x55555706cf90, name=name@entry=0x555555d0be97 "realized", errp=errp@entry=0x7fffffffd8c8)
    at /home/rob/src/nemu/qom/qom-qobject.c:27
#7  0x0000555555b428e0 in object_property_set_bool (obj=0x55555706c5f0, value=<optimised out>, name=0x555555d0be97 "realized", errp=0x7fffffffd8c8) at /home/rob/src/nemu/qom/object.c:1188
#8  0x0000555555953cfa in qdev_device_add (opts=0x5555569c3750, errp=0x7fffffffd9a0) at /home/rob/src/nemu/qdev-monitor.c:627
#9  0x0000555555956417 in device_init_func (opaque=<optimised out>, opts=<optimised out>, errp=<optimised out>) at /home/rob/src/nemu/vl.c:2354
#10 0x0000555555c32eea in qemu_opts_foreach (list=<optimised out>, func=0x5555559563f0 <device_init_func>, opaque=0x0, errp=0x0) at /home/rob/src/nemu/util/qemu-option.c:1105
#11 0x00005555558053e7 in main (argc=<optimised out>, argv=<optimised out>, envp=<optimised out>) at /home/rob/src/nemu/vl.c:4633
```

rbradford committed:
i386: virt: Initialise AcpiConfiguration pointer earlier

Previously the vms->acpi_configuration pointer was setup in the "done"
notification for machine initialisation. However this is too late when
memory configuration details are provided on the command line (and thus
the memory hotplug functions are used.)

As nothing in this struct relies on values that are set between
virt_machine_state_init and virt_machine_done then we can safely
initialise this this pointer and populate the structure during
virt_machine_state_init.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford commented:
@yangzhon I was able to identify the crash and pushed a fix to this branch. It no longer crashes for me, lets see if it clears Semaphore. Can you test memory hotplug with my change?

yangzhon commented:
@rbradford,

Thanks very much for your fix!!!

Yes, you are right! Since the CI's nemu command has nvdimm device, which hotplug the nvdimm before virt_machine_done() is called.  Our qemu commands do not have nvdimm device and when we do the memory hotplug in monitor, the  virt_machine_done()  has been called. So we could not find this issue.

As for conf->hotplug_memory.base, which will be only used in acpi table build and build_srat() will use this. So, in my patch, virt_dimm_plug() and virt_dimm_unplug(), i will change conf->hotplug_memory.base to vms->hotplug_memory.base. I will only change this and please help review again, thanks!

yangzhon commented:
@sameo @mcastelino @rbradford 

I re-send this PR with rob's patch, please help review.

I have verified the memory hotplug in this new patch set, and memory hotplug work well.  thanks!

yangzhon commented:
@sameo @mcastelino @rbradford

If this PR is not any issue, please merge this PR. Sebastien will do nvdimm porting based on this PR, In my side, build_srat acpi table need support as #35.   thanks!
